### PR TITLE
CAS Client upgrade to 3.4.x

### PR DESCRIPTION
**Pull request by mmoayyed**
Closes https://github.com/pac4j/pac4j/issues/213 and https://github.com/pac4j/pac4j/issues/242 and https://github.com/pac4j/pac4j/issues/222

### CODE REVIEW

File: pac4j-cas/src/main/java/org/pac4j/cas/client/CasClient.java [hunk 1]
```diff
@@ -150,6 +154,87 @@ protected void internalInit() {
         if (CommonHelper.isBlank(this.casLoginUrl) && CommonHelper.isBlank(this.casPrefixUrl)) {
             throw new TechnicalException("casLoginUrl and casPrefixUrl cannot be both blank");
         }
+
+        initializeClientConfiguration();
+
+        if (this.casProtocol == CasProtocol.CAS10) {
```
Comment (mmoayyed) [inline pac4j-cas/src/main/java/org/pac4j/cas/client/CasClient.java]: I reorganized this block to smaller methods.

File: pac4j-cas/src/main/java/org/pac4j/cas/client/CasClient.java [hunk 2]
```diff
@@ -202,20 +261,23 @@ protected CasCredentials retrieveCredentials(final WebContext context) throws Re
             final CasCredentials casCredentials = new CasCredentials(ticket, getName());
             logger.debug("casCredentials : {}", casCredentials);
             return casCredentials;
-        } else if (this.logoutHandler.isLogoutRequest(context)) {
+        }
+
+        if (this.logoutHandler.isLogoutRequest(context)) {
```
Comment (mmoayyed) [inline pac4j-cas/src/main/java/org/pac4j/cas/client/CasClient.java]: Took out else clauses that were not needed. A return at the end of every block will terminate execution.

File: pac4j-cas/src/main/java/org/pac4j/cas/client/rest/AbstractCasRestClient.java [hunk 3]
```diff
@@ -70,7 +71,7 @@ public void destroyTicketGrantingTicket(final WebContext context, final HttpTGTP
             final URL deleteURL = new URL(endpointURL, endpointURL.getPath() + "/" + profile.getTicketGrantingTicketId());
             connection = HttpUtils.openDeleteConnection(deleteURL);
             final int responseCode = connection.getResponseCode();
-            if (responseCode != HttpStatus.SC_OK) {
+            if (responseCode != HTTP_STATUS_OK) {
```
Comment (mmoayyed) [inline pac4j-cas/src/main/java/org/pac4j/cas/client/rest/AbstractCasRestClient.java]: We were pulling in httpclient transitively before. I created constants to remove that dependency.

File: pac4j-cas/src/main/java/org/pac4j/cas/client/CasClient.java [hunk 4]
```diff
@@ -93,11 +95,13 @@
...
+    protected String encoding = "UTF-8";
+
     protected String casLoginUrl;
 
     protected String casPrefixUrl;
     
-    protected long timeTolerance=1000L;
+    protected long timeTolerance = 1000L;
 
     protected CasProtocol casProtocol = CasProtocol.CAS20;
```
Comment (leleuj) [inline pac4j-cas/src/main/java/org/pac4j/cas/client/CasClient.java]: I think we should use the CAS protocol version 3.0 by default, shouldn't we?

File: pac4j-cas/src/main/java/org/pac4j/cas/client/rest/AbstractCasRestClient.java (hunk 3 above)
Comment (leleuj) [inline pac4j-cas/src/main/java/org/pac4j/cas/client/rest/AbstractCasRestClient.java]: You already have this constant here: https://github.com/pac4j/pac4j/blob/master/pac4j-core/src/main/java/org/pac4j/core/context/HttpConstants.java

File: pac4j-cas/src/main/java/org/pac4j/cas/client/rest/CasRestAuthenticator.java [hunk 5]
```diff
@@ -40,6 +39,8 @@
     private final String casServerPrefixUrl;
     private final String casRestUrl;
 
+    private static final int HTTP_STATUS_CREATED = 201;
```
Comment (leleuj) [inline pac4j-cas/src/main/java/org/pac4j/cas/client/rest/CasRestAuthenticator.java]: This one should be add in the `HttpConstants` interface.

**Comment by leleuj** [comment]
Excellent!

+1 with a few comments

**Comment by mmoayyed** [comment]
Updated pull

File: pac4j-core/src/main/java/org/pac4j/core/context/HttpConstants.java [hunk 6]
```diff
@@ -23,23 +23,24 @@
...
-    public static final String LOCATION_HEADER = "Location";
+    int DEFAULT_PORT = 80;
 
-    public static final String AUTHORIZATION_HEADER = "Authorization";
+    String LOCATION_HEADER = "Location";
 
-    public static final String AUTHENTICATE_HEADER = "WWW-Authenticate";
+    String AUTHORIZATION_HEADER = "Authorization";
 
-    public static final String CONTENT_TYPE_HEADER = "Content-Type";
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/context/HttpConstants.java]: Why did you remove these constants?

Comment (mmoayyed) [inline pac4j-core/src/main/java/org/pac4j/core/context/HttpConstants.java]: Strange. This may have been the IDE's fault. Let me check.

Comment (mmoayyed) [inline pac4j-core/src/main/java/org/pac4j/core/context/HttpConstants.java]: OK I see. IDEA did the following:

1. Removed `public static final` as that is redundant.
2. Remove unused constants. Looks like there were 2 that were not used anywhere?

Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/context/HttpConstants.java]: Yes, indeed, not in `pac4j`, but I've seen them used in the `play-pac4j` implementation (at least).

Comment (mmoayyed) [inline pac4j-core/src/main/java/org/pac4j/core/context/HttpConstants.java]: OK. should these constants be moved to the appropriate projects then? or are you saying there may be other projects that could use these? I can restore them, and disable inspection clean up in IDE.

Comment (mmoayyed) [inline pac4j-core/src/main/java/org/pac4j/core/context/HttpConstants.java]: Restored

Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/context/HttpConstants.java]: I think they might be used in many implementations. If ever that becomes false, we'll deprecate and finally remove them.

**Comment by leleuj** [comment]
Great! Thanks
